### PR TITLE
Trainer module implemented

### PR DESCRIPTION
**Pull request by domino468**

### CODE REVIEW

File: src/main/java/com/coursesPlatform/coursePortfolio/Trainer.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package com.coursesPlatform.coursePortfolio;
+
+import javax.persistence.*;
+
+@Entity
+public class Trainer {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/Trainer.java:6]: can be non public?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerController.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/trainer")
+@CrossOrigin(origins = "http://localhost:4200/")
+ class TrainerController {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerController.java:12]: format it please

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerDTO.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public TrainerDTO(String name, String lastName, String mail, String phoneNumber, String biography) {
+        this.name = name;
+        this.lastName = lastName;
+        this.mail = mail;
+        this.phoneNumber = phoneNumber;
+        this.biography = biography;
+    }
+
+    public TrainerDTO(String name, String lastName) {
+
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerDTO.java:21]: is it needed?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerDTO.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
+package com.coursesPlatform.coursePortfolio;
+
+import java.util.Objects;
+
+public class TrainerDTO {
+    private String name;
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerDTO.java:6]: why not private final?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerExternalDTO.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package com.coursesPlatform.coursePortfolio;
+
+import java.util.List;
+import java.util.Objects;
+
+public class TrainerExternalDTO {
+    private final String name;
+    private final  String lastName;
+    private String biography;
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerExternalDTO.java:9]: why not final?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerExternalDTO.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    public String getLastName() {
+        return lastName;
+    }
+
+    public String getBiography() {
+        return biography;
+    }
+
+    public void setBiography(String biography) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerExternalDTO.java:29]: why setter?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerInternalDTO.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
+package com.coursesPlatform.coursePortfolio;
+
+import java.util.Objects;
+
+public class TrainerInternalDTO {
+    private final String name;
+    private final String lastName;
+    private String  mail;
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerInternalDTO.java:8]: why those non-final fields?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package com.coursesPlatform.coursePortfolio;
+
+public class TrainerNotFoundException extends  RuntimeException{
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerNotFoundException.java:3]: format it please

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.coursesPlatform.coursePortfolio;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+public interface TrainerRepository extends JpaRepository<Trainer , Long> {
+    Optional<Trainer> findByNameAndLastName(String name, String lastName);
+
+
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerRepository.java:10]: empty lines...

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+@Service
+public class TrainerService {
+    private final TrainerRepository trainerRepository;
+
+    public TrainerService(TrainerRepository trainerRepository) {
+        this.trainerRepository = trainerRepository;
+    }
+
+    TrainerDTO convertTrainerToDTO(Trainer trainer) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:17]: can be private, should be lower in this file

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        return new TrainerDTO(
+                trainer.getName(),
+                trainer.getLastName(),
+                trainer.getMail(),
+                trainer.getPhoneNumber(),
+                trainer.getBiography()
+        );
+    }
+
+    Trainer convertDTOToTrainer(TrainerDTO trainerDTO) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:27]: can be private, should be lower in this file

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        trainer.setLastName(trainerDTO.getLastName());
+        trainer.setMail(trainerDTO.getMail());
+        trainer.setPhoneNumber(trainerDTO.getPhoneNumber());
+        trainer.setBiography(trainerDTO.getBiography());
+
+        return trainer;
+
+    }
+
+    TrainerExternalDTO convertTrainerToExternalDTO(Trainer trainer) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:39]: can be private, should be lower in this file

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        return new TrainerExternalDTO(
+                trainer.getName(),
+                trainer.getLastName(),
+                trainer.getBiography()
+        );
+    }
+
+    TrainerInternalDTO convertTrainerToInternalDTO(Trainer trainer) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:47]: can be private, should be lower in this file

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerInternalDTO.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
+package com.coursesPlatform.coursePortfolio;
+
+import java.util.Objects;
+
+public class TrainerInternalDTO {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerInternalDTO.java:5]: Is there any difference between TrainerDTO and TrainerInternalDTO?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    void delete(TrainerDTO trainerDTO) {
+        trainerRepository
+                .findByNameAndLastName(trainerDTO.getName(), trainerDTO.getLastName())
+        .orElseThrow(TrainerNotFoundException::new);
+        trainerRepository.deleteByNameAndLastName(trainerDTO.getName(),trainerDTO.getLastName());
+    }
+
+    void deleteByNameAndLastName(String name, String lastName) {
```
Comment (LukaszGryziewicz) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:103]: check if trainer exists please

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    TrainerDTO update(TrainerDTO updatedTrainer) {
+        Optional<Trainer> existingTrainer = trainerRepository
+                .findByNameAndLastName(updatedTrainer.getName(), updatedTrainer.getLastName());
+
+        Trainer trainer = existingTrainer.orElseThrow(TrainerNotFoundException::new);
+        trainer.update(convertDTOToTrainer(updatedTrainer));
+        return updatedTrainer;
+    }
+
+    void delete(TrainerDTO trainerDTO) {
```
Comment (LukaszGryziewicz) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:96]: remove this method, the one below is better

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        return new TrainerInternalDTO(
+                trainer.getName(),
+                trainer.getLastName(),
+                trainer.getMail(),
+                trainer.getPhoneNumber(),
+                trainer.getBiography()
+        );
+    }
+
+    List<TrainerDTO> findAllTrainers() {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:57]: Inconsistent naming - find or show?

File: src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+                .stream()
+                .map(this::convertTrainerToInternalDTO)
+                .collect(Collectors.toList());
+    }
+
+    TrainerDTO add(TrainerDTO trainerDTO) {
+
+        Optional<Trainer> titleAndDescription = trainerRepository
+                .findByNameAndLastName(trainerDTO.getName(), trainerDTO.getLastName());
+
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/TrainerService.java:81]: empty lines

File: src/test/java/com/coursesPlatform/coursePortfolio/TrainerControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,4 @@
+package com.coursesPlatform.coursePortfolio;
+
+public class TrainerControllerTest {
```
Comment (adamkulawik) [inline src/test/java/com/coursesPlatform/coursePortfolio/TrainerControllerTest.java:3]: seems to be unnecessary

File: src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+@Transactional
+public class TrainerServiceTest {
+    @Autowired
+    private TrainerService trainerService;
+    @Autowired
+    private TrainerRepository trainerRepository;
+
```
Comment (adamkulawik) [inline src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java:21]: empty lines?

File: src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    void shouldAddTrainer() {
+        //given
+        TrainerDTO trainer = new TrainerDTO("ada", "ada");
+        //when
+        trainerService.add(trainer);
+        //then
+        assertThat(trainerService.findAllTrainers()).containsExactlyInAnyOrder(trainer);
+    }
```
Comment (adamkulawik) [inline src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java:33]: no empty line - check all methods

File: src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+        //when
+        TrainerDTO update = trainerService.update(newTrainer);
+        //then
+        TrainerDTO byNameAndLastName = trainerService.findByNameAndLastName("Adam", "Dominik");
+        assertThat(byNameAndLastName).isEqualTo(newTrainer);
+    }
+}
+
+
```
Comment (adamkulawik) [inline src/test/java/com/coursesPlatform/coursePortfolio/TrainerServiceTest.java:70]: empty lines

**Review by adamkulawik — CHANGES_REQUESTED** [review]

**Review by LukaszGryziewicz — COMMENTED** [review]
Please remove empty lines

File: src/main/java/com/coursesPlatform/coursePortfolio/Trainer.java [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+    private String name;
+    private String lastName;
+    private String  mail;
+    private String phoneNumber;
+    private String biography;
+
+    public Trainer() {
+    }
+
+    public Trainer( String name, String lastName, String mail, String phoneNumber, String biography) {
```
Comment (adamkulawik) [inline src/main/java/com/coursesPlatform/coursePortfolio/Trainer.java:19]: why all those public methods?

**Review by adamkulawik — APPROVED** [review]

**Pull request merged**